Sources/DistributedActors/Plugins/ClusterSingleton/ClusterSingleton.swift

```diff
@@ -0,0 +1,448 @@
+        // TODO: perhaps we can figure out where `to` is next and hand over gracefully?
+//        self.handOver(to: nil)
+        self.clusterEventsSubscribeTask?.cancel()
+    }
+
+    private func receiveClusterEvent(_ event: Cluster.Event) async throws {
+        // Feed the event to `AllocationStrategy` then forward the result to `updateTargetNode`,
+        // which will determine if `targetNode` has changed and react accordingly.
+        let node = self.allocationStrategy.onClusterEvent(event)
+        try await self.updateTargetNode(node: node)
+    }
+
+    private func updateTargetNode(node: UniqueNode?) async throws {
+        guard self.targetNode != node else {
+            self.log.debug("Skip updating target node. New node is already the same as current targetNode.", metadata: self.metadata())
+            return
+        }
+
+        let selfNode = self.actorSystem.cluster.uniqueNode
+
+        let previousTargetNode = self.targetNode
+        self.targetNode = node
+
+        switch node {
+        case selfNode:
+            try await self.takeOver(from: previousTargetNode)
+        default:
+            if previousTargetNode == selfNode {
+                self.handOver(to: node)
+            }
+
+            // Update `singleton` regardless
+            try self.updateSingleton(node: node)
+        }
+    }
+
+    private func takeOver(from: UniqueNode?) async throws {
+        guard let singletonFactory = self.singletonFactory else {
+            preconditionFailure("Cluster singleton [\(self.settings.name)] cannot run on this node. Please review AllocationStrategySettings and/or cluster singleton usage.")
+        }
+
+        self.log.debug("Take over singleton [\(self.settings.name)] from [\(String(describing: from))]", metadata: self.metadata())
+
+        // TODO: (optimization) tell `from` node that this node is taking over (https://github.com/apple/swift-distributed-actors/issues/329)
+        let singleton = try await _Props.$forSpawn.withValue(_Props.singleton(settings: self.settings)) {
+            try await singletonFactory(self.actorSystem)
+        }
+        self.updateSingleton(singleton)
+    }
+
+    private func handOver(to: UniqueNode?) {
+        self.log.debug("Hand over singleton [\(self.settings.name)] to [\(String(describing: to))]", metadata: self.metadata())
+
+        // TODO: (optimization) tell `to` node that this node is handing off (https://github.com/apple/swift-distributed-actors/issues/329)
+        self.updateSingleton(nil)
+    }
+
+    private func updateSingleton(node: UniqueNode?) throws {
+        switch node {
+        case .some(let node) where node == self.actorSystem.cluster.uniqueNode:
+            ()
+        case .some(let node):
+            let singleton = try Act.resolve(id: .singleton(Act.self, settings: self.settings, remote: node), using: self.actorSystem)
```

Comment:
Just FYI, That's one of the places I have new impl based on well-known metadata for :)